source/common/common/mem_block_builder.h

```diff
@@ -0,0 +1,136 @@
+// data structures. The intended usage model is to release the raw assembled
+// memory block from the MemBlockBuilder for efficient storage.
+//
+// The goal for this class is to provide a usage model to replace direct usage
+// of memcpy with a pattern that is easy to validate for correctness by
+// inspection, asserts, and fuzzing, but when compiled for optimization is
+// roughly as efficient as raw memcpy.
+template <class T> class MemBlockBuilder {
+public:
+  // Constructs a MemBlockBuilder allowing for 'capacity' instances of T.
+  explicit MemBlockBuilder(uint64_t capacity)
+      : data_(std::make_unique<T[]>(capacity)), capacity_(capacity), capacity_remaining_(capacity),
+        write_ptr_(data_.get()) {}
+  MemBlockBuilder() : capacity_(0), capacity_remaining_(0), write_ptr_(nullptr) {}
+
+  /**
+   * Populates (or repopulates) the MemBlockBuilder to make it the specified
+   * capacity. This does not have resize semantics; when populate() is called any
+   * previous contents are erased.
+   *
+   * @param capacity The number of memory elements to allocate.
+   */
+  void populate(uint64_t capacity) {
+    data_ = std::make_unique<T[]>(capacity);
+    capacity_ = capacity;
+    capacity_remaining_ = capacity;
+    write_ptr_ = data_.get();
+  }
+
+  /**
+   * @return the capacity.
+   */
+  uint64_t capacity() const { return capacity_; }
+
+  /**
+   * Appends a single object of type T, moving an internal write-pointer
+   * forward. Asserts that there is room to write the object when compiled
+   * for debug.
+   *
+   * @param object the object to append.
+   */
+  void appendOne(T object) {
+    RELEASE_ASSERT(capacity_remaining_ >= 1, "insufficient capacity");
+    ASSERT(write_ptr_ < (data_.get() + capacity_));
+    *write_ptr_++ = object;
+    --capacity_remaining_;
+  }
+
+  /**
+   * Appends raw data, moving an internal write-pointer forward. Asserts
+   * that there is room to write the block when compiled for debug. It is
+   * the caller's responsibility to ensure that the input data is valid.
+   *
+   * @param data The block of objects to insert.
+   * @param size The number of elements in the block.
+   */
+  void appendData(const T* data, uint64_t size) {
+    RELEASE_ASSERT(capacity_remaining_ >= size, "insufficient capacity");
+    ASSERT((write_ptr_ + size) <= (data_.get() + capacity_));
+    if (size != 0) {
+      memcpy(write_ptr_, data, size * sizeof(T));
+    }
+    write_ptr_ += size;
+    capacity_remaining_ -= size;
+  }
+
+  /**
+   * Appends the contents of another memory block to this one.
+   *
+   * @param src the block to append.
+   */
+  void appendBlock(const MemBlockBuilder& src) { appendData(src.data_.get(), src.size()); }
+
+  /**
+   * @return the number of elements remaining in the MemBlockBuilder.
+   */
+  uint64_t capacityRemaining() const { return capacity_remaining_; }
+
+  /**
+   * Empties the contents of this.
+   */
+  void reset() {
+    data_.reset();
+    capacity_ = 0;
```

Comment:
perf nit: technically couldn't you compute capacity on the fly from capacity_remaining?

Reply:
Yes, that's a good plan as capacity() was only used in asserts in production. Now those asserts are no longer sensible as there is no longer any redundancy within the data structure to check.